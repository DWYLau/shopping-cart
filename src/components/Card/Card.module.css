.card {
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  align-items: center;
  min-height: 52vh;
  width: 85%;
  background-color: #f8f8f4;
  border: 1px solid #dcdcdc;
  border-radius: 5px;
}

.card h3 {
  text-align: center;
  width: 70%;
  font-family: system-ui, sans-serif;
  font-size: 1.1rem;
  font-weight: bold;
  line-height: 1.1;
}

.image {
  background-color: white;
  padding: 1rem;
  height: 25vh;
  width: 10vw;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
}

.rating-container {
  display: flex;
  gap: 1rem;
}

.rating,
.likes {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 0.3rem;
}

.icon {
  height: 1rem;
}

.button-container {
  display: flex;
  gap: 0.5rem;
  margin-left: 2rem;
}

.cart-button {
  cursor: pointer;
  padding: 0.7rem;
  font-family: system-ui, sans-serif;
  font-weight: bold;
  line-height: 1.1;
  border: 2px solid #f8f8f4;
  border-radius: 8px;
  background-color: #ccff00;
  min-height: 100%;
}

.cart-button:hover {
  border: 2px solid #5f9ea0;
}

.input-number {
  border: none;
  background-color: #d3d3d3;
  border: 2px solid #f8f8f4;
  border-radius: 8px;
  font-size: 1rem;
  padding: 0.4rem;
  min-height: 100%;
  width: 23%;
}

.input-number:hover {
  border: 2px solid #5f9ea0;
  background-color: #b2beb5;
}
