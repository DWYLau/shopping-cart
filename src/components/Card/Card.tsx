import styles from "./Card.module.css"
import star from "../../assets/icons/star.png"
import like from "../../assets/icons/like.png"
import { Product } from "../../utils/types"

interface Props {
  product: Product
  handleQuantity: (event: React.ChangeEvent<HTMLInputElement>) => void
  addCart: (product: Product) => void
}

function Card({ product, handleQuantity, addCart }: Props) {
  return (
    <div key={product.id} className={styles.card}>
      <img className={styles.image} src={product.image} alt='' />
      <h3>{product.title}</h3>

      <div className={styles["rating-container"]}>
        <div className={styles.rating}>
          <img className={styles.icon} src={star} alt='Rating' />
          <h3>{product.rating.rate}</h3>
        </div>

        <div className={styles.likes}>
          <img className={styles.icon} src={like} alt='Likes' />
          <h3>{product.rating.count}</h3>
        </div>

        <h3>Â£{product.price}</h3>
      </div>

      <div className={styles["button-container"]}>
        <input
          onChange={handleQuantity}
          type='number'
          min={0}
          max={100}
          defaultValue={0}
          className={styles["input-number"]}
        />
        <button
          onClick={() => addCart(product)}
          className={styles["cart-button"]}
        >
          Add To Cart
        </button>
      </div>
    </div>
  )
}

export default Card
