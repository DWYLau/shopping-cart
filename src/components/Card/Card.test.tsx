import { describe, it, expect } from "vitest"
import { BrowserRouter } from "react-router-dom"
import { getByText, render, screen } from "@testing-library/react"
import { vi } from "vitest"
import Card from "./Card"
import { product } from "./data.ts"
import userEvent from "@testing-library/user-event"

describe("Card component", () => {
  const mockHandleQuantity = vi.fn()
  const mockAddCart = vi.fn()

  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <Card
          key={product.id}
          product={product}
          handleQuantity={mockHandleQuantity}
          addCart={mockAddCart}
        />
      </BrowserRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it("renders product's title", () => {
    const { container } = render(
      <BrowserRouter>
        <Card
          key={product.id}
          product={product}
          handleQuantity={mockHandleQuantity}
          addCart={mockAddCart}
        />
      </BrowserRouter>
    )

    getByText(
      container,
      "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
    )
  })

  it("should change input value if value is typed in", async () => {
    const user = userEvent.setup()

    render(
      <BrowserRouter>
        <Card
          key={product.id}
          product={product}
          handleQuantity={mockHandleQuantity}
          addCart={mockAddCart}
        />
      </BrowserRouter>
    )

    const input = screen.getByRole("spinbutton", {
      name: "",
    }) as HTMLInputElement

    await user.type(input, "5")

    expect(input.id).toBe("quantity-input")
    expect(input.value).toBe("5")
  })
})
