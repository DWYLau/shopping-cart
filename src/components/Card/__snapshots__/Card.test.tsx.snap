// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > renders correctly 1`] = `
<div>
  <div
    class="_card_1f2933"
  >
    <img
      alt=""
      class="_image_1f2933"
      src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
    />
    <h3>
      Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
    </h3>
    <div
      class="_rating-container_1f2933"
    >
      <div
        class="_rating_1f2933"
      >
        <img
          alt="Rating"
          class="_icon_1f2933"
          src="/src/assets/icons/star.png"
        />
        <h3>
          3.9
        </h3>
      </div>
      <div
        class="_likes_1f2933"
      >
        <img
          alt="Likes"
          class="_icon_1f2933"
          src="/src/assets/icons/like.png"
        />
        <h3>
          120
        </h3>
      </div>
      <h3>
        Â£
        109.95
      </h3>
    </div>
    <div
      class="_button-container_1f2933"
    >
      <input
        class="_input-number_1f2933"
        id="quantity-input"
        max="100"
        min="0"
        type="number"
        value="0"
      />
      <button
        class="_cart-button_1f2933"
      >
        Add To Cart
      </button>
    </div>
  </div>
</div>
`;
